llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp

```diff
@@ -0,0 +1,199 @@
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+#include <string>
+
+using namespace llvm;
+
+namespace {
+
+struct CustomInliningPass : public PassInfoMixin<CustomInliningPass> {
+  PreservedAnalyses run(Function &F, FunctionAnalysisManager &AM) {
+    std::vector<CallInst *> callsToInline = findCallsToInline(F);
+
+    if (callsToInline.empty())
+      return PreservedAnalyses::all();
+
+    processCallSites(F, callsToInline);
+
+    return PreservedAnalyses::none();
+  }
+
+  static bool isRequired() { return true; }
+
+private:
+  std::vector<CallInst *> findCallsToInline(Function &F) {
+    Function *Callee{};

```

Comment:
Do we need to define `Callee` variable in the beginning of the function?

Reply:
Yes, I think that would be appropriate here. If you don't create a Callee variable, you'll have to call CI->getCalledFunction() twice, which may be less efficient.

Reply:
I'm not sure why. On the line 38 ```cpp
Callee = CI->getCalledFunction();
```
You do that unconditionally, so you override `Callee` variable on each iteration. What do you mean by calling `CI->getCalledFunction();` twice?

Reply:
I don't quite understand what your question is. As far as I understand, you can write it in different ways (a kind of syntactic sugar). You can do this function without an explicit `Callee` declaration at the beginning of the function. If I misunderstand something, please explain the problem in more detail.

```cpp
  std::vector<CallInst *> findCallsToInline(Function &F) {
    Function *Callee{};
    std::vector<CallInst *> callsToInline{};
    for (auto &BB : F) {
      for (auto &I : BB) {
        if (auto *CI = dyn_cast<CallInst>(&I)) {
          Callee = CI->getCalledFunction();
          if (Callee && Callee->arg_size() == 0 &&
              Callee->getReturnType()->isVoidTy()) {
            callsToInline.push_back(CI);
          }
        }
      }
    }
    return callsToInline;
  }
```

Reply:
Yes, you can use it like in the code snippet you shared in the previous message, or just move `Function *Callee{};` inside `if (auto *CI = dyn_cast<CallInst>(&I)) {`.

My point is that there is no need to initialize `Callee` var in outer scope